test/test_model.py

```diff
@@ -981,6 +981,52 @@ def test_juju_log(self):
             self.backend.juju_log('BAR', 'foo')
         self.assertEqual(fake_script_calls(self, clear=True), [['juju-log', '--log-level', 'BAR', 'foo']])
 
+    def test_metrics(self):
+        fake_script(self, 'add-metric', 'exit 0')
+        test_cases = [(
+            {'foo': 42, 'b-ar': 4.5, 'ba_-z': 4.5, 'a': 1},
+            {'de': 'ad', 'be': 'ef_ -'},
+            [['add-metric', '--labels', 'de=ad,be=ef_ -', 'foo=42', 'b-ar=4.5', 'ba_-z=4.5', 'a=1']]
+        ), (
+            {'foo1': 0, 'b2r': 4.5},
+            {'d3': 'aд', 'b33f': '3_ -'},
+            [['add-metric', '--labels', 'd3=aд,b33f=3_ -', 'foo1=0', 'b2r=4.5']],
+        )]
+        for metrics, labels, expected_calls in test_cases:
+            self.backend.add_metrics(metrics, labels)
+            self.assertEqual(fake_script_calls(self, clear=True), expected_calls)
+
+        invalid_inputs = [
```

Comment:
I would rather see several test cases, one for the happy behavior, and one for the 'these are invalid'. I'd feel better with a separate test case each for invalid metric name, metric values, label names, label values.

Reply:
https://github.com/canonical/operator/pull/140/commits/09f96263967a27147330526271538e93c166381e